### PR TITLE
Towards TF dataset, part I

### PR DESCRIPTION
**Pull request by jindrahelcl**
* Lazifying graph construction
* New abstraction scheme (graph executors and executable nested classes)
* Tensor Runner changed to operate indirectly 
* New configuration disambiguation module
* Training profiler class

### CODE REVIEW

File: neuralmonkey/config/disambiguate.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
+"""Module for disambiguating and enhancing configuration."""
```
Comment (jlibovicky) [inline neuralmonkey/config/disambiguate.py]: To není _disambiguate_, protože ta konfigurace není ambiguious. Podle mě je to spíš něco jako _resolving parameter values_.

File: neuralmonkey/experiment.py [hunk 2]
```diff
@@ -105,6 +106,21 @@ def model(self) -> Namespace:
 
         return self._model
 
+    def _bless_graph_executors(self) -> None:
```
Comment (jlibovicky) [inline neuralmonkey/experiment.py]: To chce docstring, neznalý naší teologie netuší, co je to žehnat.

File: neuralmonkey/runners/base_runner.py [hunk 3]
```diff
@@ -34,41 +39,106 @@ class ExecutionResult(NamedTuple(
     """
 
 
-class Executable:
+class GraphExecutor(GenericModelPart):
```
Comment (jlibovicky) [inline neuralmonkey/runners/base_runner.py]: Docstring

File: neuralmonkey/runners/base_runner.py [hunk 4]
```diff
@@ -34,41 +39,106 @@ class ExecutionResult(NamedTuple(
...
 
-    @property
-    def result(self) -> Optional[ExecutionResult]:
-        return getattr(self, "_result")
+    class Executable(Generic[Executor]):
```
Comment (jlibovicky) [inline neuralmonkey/runners/base_runner.py]: Docstring (včetně toho, proč je to takhle divně generický)

File: neuralmonkey/config/disambiguate.py (hunk 1 above)
Comment (jindrahelcl) [reply neuralmonkey/config/disambiguate.py]: a jak bys pojmenoval ten modul? mě to jako disambiguace přijde, když tam můžu mít třeba val_dataset nebo trainers jako seznam i jako jednu hodnotu.

**Comment by jindrahelcl** [comment]
Added:

* Objective refactor (lazy cost)
* Learning utils' training loop function take the whole config instead of million parameters
* TF-based vocabulary lookup function
* Dataset runner
* Big ordnung in tests

File: neuralmonkey/tf_manager.py [hunk 5]
```diff
@@ -236,16 +229,17 @@ def execute(self,
             A list of `ExecutionResult` tuples, one for each executable
             (runner).
         """
+        feedables = set.union(*[runner.feedables for runner in runners])
+        default_feed_dict = _feed_dicts(batch, feedables, train=train)
```
Comment (jlibovicky) [inline neuralmonkey/tf_manager.py]: Takže jsi zahodil ty additional feed dicty?

File: neuralmonkey/attention/feed_forward.py [hunk 6]
```diff
@@ -170,6 +166,11 @@ def attention(self,
         return context, next_loop_state
 
     def initial_loop_state(self) -> AttentionLoopState:
+        # TODO blessing
```
Comment (jlibovicky) [inline neuralmonkey/attention/feed_forward.py]: Co přesně myslíš tím `TODO` jako že tam chceš napsat, proč se to děje?

File: neuralmonkey/runners/logits_runner.py [hunk 7]
```diff
@@ -100,36 +83,25 @@ def __init__(self,
             raise ValueError("Either a pick index or a vocabulary value can "
                              "be specified, not both at the same time.")
 
-        self._pick_index = None  # type: Optional[int]
+        self.pick_index = None  # type: Optional[int]
```
Comment (jlibovicky) [inline neuralmonkey/runners/logits_runner.py]: (není čas přejít na doopravické anotace z Pythonu 3.6?)

File: neuralmonkey/decoders/transformer.py [hunk 8]
```diff
@@ -161,33 +161,10 @@ def __init__(self,
...
-        else:
-            self.dimension = self.embedding_size
-
-        if not self.dimension:
-            raise ValueError("Decoder could not infer model dimension")
-
-        if self.embedding_size != self.dimension:
-            raise ValueError("Model dimension and input embedding size"
-                             "do not match")
+        self.encoder_states = lambda: [get_attention_states(e)
```
Comment (jlibovicky) [inline neuralmonkey/decoders/transformer.py]: Proč jsou to lambdy a ne `@tensor`?

File: neuralmonkey/config/disambiguate.py (hunk 1 above)
Comment (jlibovicky) [reply neuralmonkey/config/disambiguate.py]: Normalize nebo standardize

File: neuralmonkey/config/disambiguate.py [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+    elif delayed_trainers:
+        assert len(cfg.trainers) == 1
+        denominator = cfg.trainers[0].batches_per_update
+
+    cfg.log_timer = _resolve_period(cfg.logging_period, denominator)
+    cfg.val_timer = _resolve_period(cfg.validation_period, denominator)
+
+
+def _resolve_period(period: Union[str, int],
+                    denominator: int) -> Callable[[int, float], bool]:
```
Comment (jlibovicky) [inline neuralmonkey/config/disambiguate.py]: Docstring, že to resovluje to vrací něco, co rozhoduje, jestli se má logovat trénovaání nebo validace.

File: neuralmonkey/runners/beamsearch_runner.py [hunk 10]
```diff
@@ -17,13 +16,11 @@
 class BeamSearchExecutable(Executable):
     def __init__(self,
                  rank: int,
```
[the remark this answers is not in the document]
Comment (jindrahelcl) [reply neuralmonkey/runners/beamsearch_runner.py]: ne, tadyten executable je pro jeden konkrétní rank v beamu. Proto tam máme tu runners range, která ti udělá tolik runnerů, kolik je beam size

File: neuralmonkey/trainers/generic_trainer.py [hunk 11]
```diff
@@ -253,7 +253,7 @@ def next_to_execute(self) -> NextExecute:
         fetches["losses"] = self.trainer.objective_values
         fetches["_update_ops"] = tf.get_collection(tf.GraphKeys.UPDATE_OPS)
 
-        return self.trainer.feedables, fetches, [{}]
+        return fetches, [{}]
```
[the remark this answers is not in the document]
Comment (jindrahelcl) [reply neuralmonkey/trainers/generic_trainer.py]: Ne, tam by asi mohl bejt jen prázdnej seznam.

File: neuralmonkey/tf_manager.py (hunk 5 above)
Comment (jindrahelcl) [reply neuralmonkey/tf_manager.py]: tadyty dva řádky jsou jen refactor, díky kterýmu `_run_executables` nemusí žrát `batch`, `feedables` a `train`.

Mění se pořadí tvorby feed dictů: Bylo to tak, že se napřed (v `_run_executables`) vygenerovaly additional feed dicty z runnerů a pak se updatovaly feed dictama z feedables. Teď je to tak, že se napřed (už v týhle funkci) vygenerujou feed dicty z feedables (těm říkejme `default`, protože v nich nejsou extroviny z executablů) a potom se zavolá funkce `_run_executables`, uvnitř který (na ř. 188 v tomhle komitu) se jako vždycky zavolají additional feed dicts.

Co executables už nevrací nejsou additional feed dicty, ale seznamy coderů, který si tu držíme vedle v tom setu feedables.

Opravil se tim mimochodem latentní bug, kterej by se projevil v tom, že kdybys plnil placeholder z executablu, přepsal by se feed dictem z model partu.

File: neuralmonkey/attention/feed_forward.py (hunk 6 above)
Comment (jindrahelcl) [reply neuralmonkey/attention/feed_forward.py]: to znamená "udělej s tim něco než se to uzavře". Mělo by tam bejt `TODO(tf-data)` a ne jen TODO.. :-)

Jednak chci dát pryč ten log a dát tam jen debug, nebo ještě něco míň nápadnýho. Plus v komentáři vysvětlit, co to tam dělá a proč to tam musí bejt. Obě dvě věci vyřeším jako response na tenhle review.

File: neuralmonkey/decoders/transformer.py (hunk 8 above)
Comment (jindrahelcl) [reply neuralmonkey/decoders/transformer.py]: to je humus kvůli beam searchi. Každej autoregresivní dekodér musí mít tohle, aby se mu místo tohohle během beam searche mohl podstrčit tensor expandovaný do batche.

Když se podíváš na tensorflowovej beam search decoder, řešej tam podobný pekla, takže bych to nechal.
Lambda tu je kvůli laziness.
Kdyby to bylo @tensor, nedalo by se to v poděděný třídě předefinovat.

File: neuralmonkey/runners/logits_runner.py (hunk 7 above)
Comment (jindrahelcl) [reply neuralmonkey/runners/logits_runner.py]: Je, ale změní se tím 39 souborů.

Comment (jindrahelcl) [reply neuralmonkey/runners/logits_runner.py]: a uprostřed branche to udělat by byla sebemrskačská orgie

File: neuralmonkey/trainers/test_multitask_trainer.py [hunk 12]
```diff
@@ -60,24 +61,25 @@ def test_mt_trainer(self):
         self.assertTrue(trainer.trainer_idx == 0)
 
         executable = trainer.get_executable()
-        mparts, fetches, feeds = executable.next_to_execute()
-        self.assertSetEqual(mparts, {self.mpart})
+        # mparts = trainer.feedables
+        fetches, feeds = executable.next_to_execute()
+        # self.assertSetEqual(mparts, {self.mpart})
```
[the remark this answers is not in the document]
Comment (jindrahelcl) [reply neuralmonkey/trainers/test_multitask_trainer.py]: tohle mazat nebudu, protože se celej ten test bude muset předělat a tohle zakomentovaný možná poslouží.

File: neuralmonkey/trainers/test_multitask_trainer.py [hunk 13]
```diff
@@ -60,24 +61,25 @@ def test_mt_trainer(self):
         self.assertTrue(trainer.trainer_idx == 0)
 
         executable = trainer.get_executable()
-        mparts, fetches, feeds = executable.next_to_execute()
-        self.assertSetEqual(mparts, {self.mpart})
+        # mparts = trainer.feedables
```
[the remark this answers is not in the document]
Comment (jindrahelcl) [reply neuralmonkey/trainers/test_multitask_trainer.py]: tohle mazat nebudu, protože se celej ten test bude muset předělat a tohle zakomentovaný možná poslouží.

File: neuralmonkey/trainers/test_multitask_trainer.py [hunk 14]
```diff
@@ -60,24 +61,25 @@ def test_mt_trainer(self):
...
         self.assertFalse(feeds[0])
 
         self.assertTrue(trainer.trainer_idx == 1)
         self.assertTrue(fetches["losses"][0] == self.mpart.loss)
 
         executable = trainer.get_executable()
-        mparts, fetches, feeds = executable.next_to_execute()
-        self.assertSetEqual(mparts, {self.mpart_2})
+        fetches, feeds = executable.next_to_execute()
+        # self.assertSetEqual(mparts, {self.mpart_2})
```
[the remark this answers is not in the document]
Comment (jindrahelcl) [reply neuralmonkey/trainers/test_multitask_trainer.py]: tohle mazat nebudu, protože se celej ten test bude muset předělat a tohle zakomentovaný možná poslouží.

File: neuralmonkey/trainers/test_multitask_trainer.py [hunk 15]
```diff
@@ -60,24 +61,25 @@ def test_mt_trainer(self):
...
         self.assertFalse(feeds[0])
 
         self.assertTrue(trainer.trainer_idx == 2)
         self.assertTrue(fetches["losses"][0] == self.mpart_2.loss)
 
         executable = trainer.get_executable()
-        mparts, fetches, feeds = executable.next_to_execute()
-        self.assertSetEqual(mparts, {self.mpart})
+        fetches, feeds = executable.next_to_execute()
+        # self.assertSetEqual(mparts, {self.mpart})
```
[the remark this answers is not in the document]
Comment (jindrahelcl) [reply neuralmonkey/trainers/test_multitask_trainer.py]: tohle mazat nebudu, protože se celej ten test bude muset předělat a tohle zakomentovaný možná poslouží.

File: neuralmonkey/trainers/generic_trainer.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (jindrahelcl) [reply neuralmonkey/trainers/generic_trainer.py]: btw. nevšiml jsem si nikde, že by se tu vracel prázdnej seznam, vždycky se vrací `[{}]`. Máš odkaz?

File: neuralmonkey/config/disambiguate.py (hunk 1 above)
Comment (jindrahelcl) [reply neuralmonkey/config/disambiguate.py]: done

File: neuralmonkey/runners/base_runner.py [hunk 16]
```diff
@@ -40,32 +43,54 @@ class Executable:
...
                        compute_losses: bool,
                        summaries: bool,
                        num_sessions: int) -> Executable:
         raise NotImplementedError()
 
+
+class BaseRunner(GraphExecutor, Generic[MP]):
+    def __init__(self,
+                 output_series: str,
+                 decoder: MP) -> None:
```
[the remark this answers is not in the document]
Comment (jindrahelcl) [reply neuralmonkey/runners/base_runner.py]: dávám to tam do TODO, protože se s tim v budoucím kódu dosti šachuje, tak bych si tam všude nasekal konflikty
